yjit_codegen.c

```diff
@@ -915,73 +917,170 @@ gen_opt_ge(jitstate_t* jit, ctx_t* ctx)
+        {
+            ctx_t reset_depth = *ctx;
+            reset_depth.chain_depth = 0;
+
+            blockid_t jump_block = { jit->iseq, jit_next_insn_idx(jit) };
+
+            // Generate the jump instruction
+            gen_direct_jump(
+                &reset_depth,
+                jump_block
+            );
+        }
+        return YJIT_END_BLOCK;
     }
+    else if (CLASS_OF(comptime_recv) == rb_cHash) {
+        if (!assume_bop_not_redefined(jit->block, HASH_REDEFINED_OP_FLAG, BOP_AREF)) {
+            return YJIT_CANT_COMPILE;
+        }
 
-    // Pop the stack operands
-    x86opnd_t idx_opnd = ctx_stack_pop(ctx, 1);
-    x86opnd_t recv_opnd = ctx_stack_pop(ctx, 1);
-    mov(cb, REG0, recv_opnd);
+        // Pop the stack operands
+        x86opnd_t idx_opnd = ctx_stack_pop(ctx, 1);
+        x86opnd_t recv_opnd = ctx_stack_pop(ctx, 1);
+        mov(cb, REG0, recv_opnd);
 
-    // if (SPECIAL_CONST_P(recv)) {
-    // Bail if it's not a heap object
-    test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK));
-    jnz_ptr(cb, side_exit);
-    cmp(cb, REG0, imm_opnd(Qfalse));
-    je_ptr(cb, side_exit);
-    cmp(cb, REG0, imm_opnd(Qnil));
-    je_ptr(cb, side_exit);
-
-    // Bail if recv has a class other than ::Array.
-    // BOP_AREF check above is only good for ::Array.
-    mov(cb, REG1, mem_opnd(64, REG0, offsetof(struct RBasic, klass)));
-    mov(cb, REG0, const_ptr_opnd((void *)rb_cArray));
-    cmp(cb, REG0, REG1);
-    jne_ptr(cb, side_exit);
+        // if (SPECIAL_CONST_P(recv)) {
+        // Bail if receiver is not a heap object
+        test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK));
+        jnz_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qfalse));
+        je_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qnil));
+        je_ptr(cb, side_exit);
 
-    // Bail if idx is not a FIXNUM
-    mov(cb, REG1, idx_opnd);
-    test(cb, REG1, imm_opnd(RUBY_FIXNUM_FLAG));
-    jz_ptr(cb, side_exit);
+        // Bail if recv has a class other than ::Hash.
+        // BOP_AREF check above is only good for ::Hash.
+        mov(cb, REG1, mem_opnd(64, REG0, offsetof(struct RBasic, klass)));
+        mov(cb, REG0, const_ptr_opnd((void *)rb_cHash));
+        cmp(cb, REG0, REG1);
+        jit_chain_guard(JCC_JNE, jit, &starting_context, OPT_AREF_MAX_CHAIN_DEPTH, side_exit);
 
-    // Save YJIT registers
-    yjit_save_regs(cb);
+        // Call VALUE rb_hash_aref(VALUE hash, VALUE key).
+        {
+            // Write incremented pc to cfp->pc as the routine can raise and allocate
+            mov(cb, REG0, const_ptr_opnd(jit->pc + insn_len(BIN(opt_aref))));
+            mov(cb, member_opnd(REG_CFP, rb_control_frame_t, pc), REG0);
+
+            // About to change REG_SP which these operands depend on. Yikes.
+            mov(cb, R8, recv_opnd);
+            mov(cb, R9, idx_opnd);
+
+            // Write sp to cfp->sp since rb_hash_aref might need to call #hash on the key
+            if (ctx->sp_offset != 0) {
```

Comment:
Maybe we should move this logic into `yjit_save_regs`. Right now we're always pushing REG_SP (twice to maintain alignment). Maybe we could always save the SP into the EC instead.

Reply:
That abstraction would break existing users that save operands that depend on REG_SP. I think it would be error prone to use. Since this is the only site at the moment I think we should wait.

Reply:
I feel like the register allocator should handle this type of things.

Reply:
That's only possible if the register allocator works as a second pass, on some kind of IR? I like the idea of abstracting things so we don't have to worry about low-level details like this all the time, but having a mini IR does have a (slight) penalty in terms of compile time, with a potential benefit in terms of maintainability and final performance of the code.

Since the register allocator could turn out to be pretty complex and a significant refactoring. I think we need some kind of mini design document before we implement the register allocator, so we can agree on a design that meets all our requirements. I outlined [some notes](https://docs.google.com/document/d/1pJADXhsxBUeZIPdjVscMKWWhjI0QnOkrn6HZuQQiJ_k/edit?usp=sharing) with the requirements I could think of. I invited you to edit so you can add another section with your ideas for a more sophisticated register allocator. If/when you feel motivated, feel free to flesh out your ideas. No deadline on this, we can both take our time to carefully weigh pros and cons, and I am genuinely open to your ideas if you think we should have a more powerful register allocator/backend.